assets/app/scripts/services/apiEndpoint_v2.js

```diff
@@ -0,0 +1,311 @@
+'use strict';
+/* jshint unused:false */
+
+// see gist: https://gist.github.com/benjaminapetersen/540ac605d3c1b0660062
+// for discussion leading up to the creation of this file
+
+angular.module('openshiftConsole')
+  .factory('ApiEndpointService_v2', function(API_CFG) {
+
+    // defaultVersion lookup for backwards compatibility...
+    var defaultVersionByGroup = {
+      'extensions':   'v1beta1',
+      'experimental': 'v1beta1',
+      '':             'v1'  // this group applies to openshift & k8s (/oapi) & (/api)
+    };
+
+    // private processor
+    var mapApis = function(groupVersions) {
+      return _.reduce(
+              groupVersions,
+              function(memo, next) {
```

Comment:
a clearer name for `memo` and `next` would help... `groups` and `groupVersion`, perhaps